src/main/java/org/elasticsearch/index/mapper/core/AbstractFieldMapper.java

```diff
@@ -481,11 +481,7 @@ public Filter termsFilter(List values, @Nullable QueryParseContext context) {
             // bit set
             return termFilter(values.get(0), context);
         default:
-            BytesRef[] bytesRefs = new BytesRef[values.size()];
-            for (int i = 0; i < bytesRefs.length; i++) {
-                bytesRefs[i] = indexedValueForSearch(values.get(i));
-            }
-            return new TermsFilter(names.indexName(), bytesRefs);
+            return new DeferredTermsFilter(names.indexName(), values, this);
```

Comment:
As you can see I also defer the copying / transformation of `values` into `BytesRef`s, which is also slow when there are many terms.